Update protocol selection page (#8276)
istio/istio.io: content/en/docs/ops/configuration/traffic-management/protocol-selection/index.md

test: no
---

Istio supports proxying any TCP traffic. This includes HTTP, HTTPS, gRPC, as well as raw TCP protocols.
In order to provide additional capabilities, such as routing and rich metrics, the protocol must be determined. This can be done automatically or explicitly specified.

Non-TCP based protocols, such as UDP, are not proxied. These protocols will continue to function as normal, without
any interception by the Istio proxy but cannot be used in proxy-only components such as ingress or egress gateways.

## Automatic protocol selection

Istio can automatically detect HTTP and HTTP/2 traffic. If the protocol cannot automatically be determined, traffic will be treated as plain TCP traffic.

> ### Server first protocols
> 
> Some protocols are "Server First" protocols, which means the server will send the first bytes. In these cases, automatic protocol detection
> will not function correctly, as Istio will determine the protocol based on the initial bytes. If you are using a server first protocol, you should
> follow the [explicit protocol selection instructions](#explicit-protocol-selection) to declare the protocol.
> 
> The following protocols are known to be incompatible with automatic protocol selection, and must also be explicitly declared or they will be treated as TCP:
> 
> |Protocol|Port|
> |--------|----|
> | SMTP   |25  |
> | DNS    |53  |
> | MySQL  |3306|
> | MongoDB|27017|
> 
> {{< tip >}}
> Since TLS communication is not server first, any TLS based traffic will support protocol detection, even if it is wrapping a protocol that does not support detection, such as MySQL.
> {{< /tip >}}

Reviewer: One more nit. The incompatible protocols part looks like it's in the `Server first protocols` subsection, so I suggest either removing the section heading or reordering the text like this:
```suggestion
The following protocols are known to be incompatible with automatic protocol selection, and must be
[explicitly declared](#explicit-protocol-selection) or they will be treated as TCP:

|Protocol|Port|
|--------|----|
| SMTP   |25  |
| DNS    |53  |
| MySQL  |3306|
| MongoDB|27017|

{{< tip >}}
Since TLS communication is not server first, any TLS based traffic will support protocol detection, even if it is wrapping a protocol that does not support detection, such as MySQL.
{{< /tip >}}

### Server first protocols

Some protocols are "Server First" protocols, which means the server will send the first bytes. In these cases, automatic protocol detection
will not function correctly, as Istio will determine the protocol based on the initial bytes. If you are using a server first protocol, you should
follow the explicit protocol selection instructions, below, to declare the protocol.
```

test: no
---

Istio supports proxying any TCP traffic. This includes HTTP, HTTPS, gRPC, as well as raw TCP protocols.
In order to provide additional capabilities, such as routing and rich metrics, the protocol must be determined. This can be done automatically or explicitly specified.

Non-TCP based protocols, such as UDP, are not proxied. These protocols will continue to function as normal, without

> any interception by the Istio proxy. They are not supported at gateways.

Reviewer: "They are not supported at gameways." is a little vague. Can you clarify? Would something like this be correct?
```suggestion
any interception by the Istio proxy but cannot be used in proxy-only components such as ingress or egress gateways.
```

Author: I think that makes sense. Basically the idea is UDP is not "supported" its just not touched at all. This means you cannot use it at gateway, but in a pod it can send UDP out untouched.

## Automatic protocol selection

Istio can automatically detect HTTP and HTTP/2 traffic. If the protocol cannot automatically be determined, traffic will be treated as plain TCP traffic.

### Server first protocols

Some protocols are "Server First" protocols, which means the server will send the first bytes. In these cases, automatic protocol detection
will not function correctly, as Istio will determine the protocol based on the initial bytes. If you are using a server first protocol, you should

> explicitly declare the protocol following the instructions below.

Reviewer: ```suggestion
follow the [explicit protocol selection instructions](#explicit-protocol-selection) to declare the protocol.
```

Author: Should it be `#manual-protocol-selection` or do I not understand how this works?

Reviewer: I thought renaming the section (manual -> explicit) would be better, since the text always uses the word explicit instead of manual. But that's just my preference.

Author: I agree

> The following protocols are known to be incompatible with automatic protocol selection, and will be treated as TCP if not explicitly declared:

Reviewer: ```suggestion
The following protocols are known to be incompatible with automatic protocol selection, and must also be explicitly declared or they will be treated as TCP:
```

|Protocol|Port|
|--------|----|
| SMTP   |25  |
| DNS    |53  |
| MySQL  |3306|
| MongoDB|27017|

{{< tip >}}
Since TLS communication is not server first, any TLS based traffic will support protocol detection, even if it is wrapping a protocol that does not support detection, such as MySQL.
{{< /tip >}}

> ## Manual protocol selection

Reviewer: ```suggestion
## Explicit protocol selection
```
